RFCXXX-PSRemoting-Providers.md

```diff
@@ -0,0 +1,298 @@
+For example the following can be used to connect to 2 hosts in parallel using the builtin SSH provider:
+
+```powershell
+Invoke-Command -HostName Host1, Host2 -ScriptBlock { 'test' }
+```
+
+Trying to implement this for custom providers would require some complicated solution like the following:
+
+```powershell
+$sessions = 'Host1', 'Host2' | ForEach-Object -Parallel {
+    New-CustomTransportSession -Target $_
+}
+
+try {
+    Invoke-Command -Session $session { 'test' }
+}
+finally {
+    $sessions | Remove-PSSession
+}
+```
+
+Introduction the `-Parallel` block itself even brings in more complications, for example `$using:var` is required when referring to vars outside of the scriptblock.
+What was a single cmdlet has now turned into multiple lines without any real error handling.
+This can be a fairly large obstacle for custom transport providers to overcome when it comes to getting people willing to try out their code vs what's built into PowerShell itself.
+
+## Proposed Solution
+
+The proposed solution would be to implement a registry of remoting providers that modules can register to.
+This registration would include a unique prefix allowing users to simply specify the transport type as part of the host connection string.
+More examples can be seen in the `User Experience` header but in short it would turn this:
+
+```powershell
+$sessions = 'Host1', 'Host2' | ForEach-Object -Parallel {
+    New-CustomTransportSession -Target $_
+}
+
+try {
+    Invoke-Command -Session $session { 'test' }
+}
+finally {
+    $sessions | Remove-PSSession
+}
+```
+
+Into
+
+```powershell
+Invoke-Command -ComputerName custom:Host1, custom:Host2
+```
+
+Existing providers can also have their own host prefix allowing things like `-ComputerName ssh:Host1` to connect over SSH to `Host1`.
+While the other parameters can't be removed straight away, it now opens the possibility to simplify the existing parameter set in the future to just a single `-ComputerName` parameter
+
+## User Experience
+
+These examples use the [PWSExecConn POC](https://github.com/jborean93/PWSExecConn) as the example custom remoting provider to see what this could potentially look like.
+To use this provider with `Invoke-Command` it would look like:
+
+```powershell
+Import-Module PWSExecConn
+
+Invoke-Command -ComputerName psexec:Hostname -ScriptBlock { 'test' }
+```
+
+Specifying connection specific options can still be done using a custom cmdlet that outputs connection specific options that inherit from `PSSessionOption` like:
+
+```powershell
+Import-Module PWSExecConn
+
+$pso = New-PWSExecSessionOption -LogonType Batch
+Invoke-Command -ComputerName psexec:Hostname -ScriptBlock { 'test' } -SessionOption $pso
+```
+
+It is also theoretically possible to embed connection specific details in the target string, for example using a username and port with SSH.
+
+```powershell
+Invoke-Command -ComputerName 'ssh:username@target:2222' -ScriptBlock { 'test' }
```

Comment:
In this example, how would the `-LogonType` be passed if it's not the default one?  The large number of parametersets and differences in capabilities/authn between transports is why we preferred transport specific cmdlets to create the session vs making the current cmdlets more complicated.  A simple URL schema registration would be ok for majority of usage that doesn't need to pass transport specific options.

Reply:
In the case where you can't specify an option in the host string, my thoughts were it should be part of the `-SessionOption` parameter, e.g.

```powershell
$pso = New-SSHSessionOption -LogonType Interactive
Invoke-Command -ComputerName ssh:hostname -SessionOption $pso
```

Each transport can have their own cmdlet to generate provider specific options which would be documented by that provider.

Reply:
Extending PSSessionOption for less likely used options seems like a good option.